### PR TITLE
add example how to make a Processing plugin

### LINKED ISSUE #3486: New plugin metadata and hooks for starting only Processing component of plugins
See https://github.com/qgis/QGIS/pull/9312

A new "hasProcessingProvider=yes" metadata tag has been added to the possible plugin metadata items, which should be set whenever a plugin implements a processing provider.

This issue is to discuss the required updates for documentation. What documentation components do you think we should include?

### PR DESCRIPTION
**Pull request by Gustry**
Fix #3486 
Linked to https://github.com/qgis/QGIS/pull/9405

CC @nyalldawson

### CODE REVIEW

File: source/docs/pyqgis_developer_cookbook/processing.rst [hunk 1]
```diff
@@ -22,18 +20,97 @@ processing interface), and a quicker development time (since Processing will tak
...
+           QgsApplication.processingRegistry().addProvider(self.provider)
+
+       def initGui(self):
+           if Qgis.QGIS_VERSION_INT < 30800:
+               self.initProcessing()
+
+       def unload(self):
+           QgsApplication.processingRegistry().removeProvider(self.provider)
+
+You can create a folder ``processing_provider`` with three files in it:
```
Comment (DelazJ) [inline source/docs/pyqgis_developer_cookbook/processing.rst]:
```suggestion
You can create a folder :file:`processing_provider` with three files in it:
```

File: source/docs/pyqgis_developer_cookbook/processing.rst [hunk 2]
```diff
@@ -22,18 +20,97 @@ processing interface), and a quicker development time (since Processing will tak
...
+
+``__init__.py`` with nothing in it. This is necessary to make a valid Python package.
```
Comment (DelazJ) [inline source/docs/pyqgis_developer_cookbook/processing.rst]: Can we use a bullet or numbered list for the three concerned files to create?
And you might want to use :file: when refering to files or folders

File: source/docs/pyqgis_developer_cookbook/processing.rst [hunk 3]
```diff
@@ -22,18 +20,97 @@ processing interface), and a quicker development time (since Processing will tak
...
 * Create a new plugin using the Plugin Builder. When the Plugin Builder asks you for
-  the template to use, select "Processing provider". 
+  the template to use, select "Processing provider".
 * The created plugin contains a provider with a single algorithm. Both the provider
   file and the algorithm file are fully commented and contain information about how to
   modify the provider and add additional algorithms. Refer to them for more information.
 
+If you want to add your existing plugin to Processing, you need to add some code.
+
+In your ``metadata.txt``, you need to add a variable:
```
Comment (DelazJ) [inline source/docs/pyqgis_developer_cookbook/processing.rst]:
```suggestion
In your :file:`metadata.txt`, you need to add a variable:
```

File: source/docs/pyqgis_developer_cookbook/processing.rst [hunk 4]
```diff
@@ -22,18 +20,97 @@ processing interface), and a quicker development time (since Processing will tak
...
+
+       def icon(self):
+           """The QIcon in the Processing toolbox."""
+           return QgsProcessingProvider.icon(self)
+
+       def loadAlgorithms(self, *args, **kwargs):
+           for alg in self.list_algorithms():
+               self.addAlgorithm(alg)
+
+Now, you can create the example algorithm. In ``example_processing_algorithm.py``,
```
Comment (DelazJ) [inline source/docs/pyqgis_developer_cookbook/processing.rst]: Maybe review this third point of the list to follow the same structure as the previous, ie sthg like
```suggestion
#. :file:`example_processing_algorithm.py`, the example algorithm file:
```

**Comment by DelazJ** [comment]
@Gustry can you also check the target issue? Should it be #3486 instead?

**Comment by Gustry** [comment]
Thanks for the review. Sorry, my first time writing documentation, not just Python ;-)

File: source/docs/pyqgis_developer_cookbook/processing.rst (hunk 2 above)
Comment (Gustry) [reply source/docs/pyqgis_developer_cookbook/processing.rst]: Done. Thanks

File: source/docs/pyqgis_developer_cookbook/processing.rst [hunk 5]
```diff
@@ -22,18 +20,97 @@ processing interface), and a quicker development time (since Processing will tak
...
+       def icon(self):
+           """The QIcon in the Processing toolbox."""
+           return QgsProcessingProvider.icon(self)
+
+       def loadAlgorithms(self, *args, **kwargs):
+           for alg in self.list_algorithms():
+               self.addAlgorithm(alg)
+
+* :file:`example_processing_algorithm.py` which contains the example algorithm file.
+  Copy/paste the content of the script template: https://github.com/qgis/QGIS/blob/master/python/plugins/processing/script/ScriptTemplate.py
```
Comment (havatv) [inline source/docs/pyqgis_developer_cookbook/processing.rst]:
```suggestion
  Copy/paste the content of the script template:
  https://github.com/qgis/QGIS/blob/master/python/plugins/processing/script/ScriptTemplate.py
```

**Comment by havatv** [comment]
Is it correct to label this with "backport release_3.4"?

**Comment by Gustry** [comment]
Yes, I would like to backport it. Even if the new metadata flag is in QGIS 3.8, the way that I'm proposing here is compatible QGIS < 3.8 too.

File: source/docs/pyqgis_developer_cookbook/processing.rst [hunk 6]
```diff
@@ -22,18 +20,98 @@ processing interface), and a quicker development time (since Processing will tak
...
+
+   from qgis.core import QgsProcessingProvider
+
+   from .example_processing_algorithm import ExampleProcessingAlgorithm
+
+
+   class Provider(QgsProcessingProvider):
+
+       @staticmethod
+       def list_algorithms(self):
```
Comment (nyalldawson) [inline source/docs/pyqgis_developer_cookbook/processing.rst]: I'm not sure on the benefit of splitting this out to a separate function - to me the example would be easier to read and understand if this was just done one by one in loadAlgorithms, e.g.

           def loadAlgorithms(self, *args, **kwargs):
                self.addAlgorithm(ExampleProcessingAlgorithm())
                # add additional algorithms here
                # self.addAlgorithm(MyOtherAlgorithm())

it's perhaps not quite as "clean", but it IS more readable and that's the main point here.

File: source/docs/pyqgis_developer_cookbook/processing.rst [hunk 7]
```diff
@@ -22,18 +20,98 @@ processing interface), and a quicker development time (since Processing will tak
...
+           """List of algorithms available in your plugin."""
+           algs = [
+               ExampleProcessingAlgorithm(),
+           ]
+           return algs
+
+       def id(self, *args, **kwargs):
+           """The ID of your plugin."""
```
Comment (nyalldawson) [inline source/docs/pyqgis_developer_cookbook/processing.rst]:
```suggestion
           """The ID of your plugin, used for identifying the provider. This string
           should be a unique, short, character only string, eg "qgis" or "gdal". This
           string should not be localised."""
```

File: source/docs/pyqgis_developer_cookbook/processing.rst [hunk 8]
```diff
@@ -22,18 +20,98 @@ processing interface), and a quicker development time (since Processing will tak
...
+           return 'yourplugin'
+
+       def name(self, *args, **kwargs):
+           """The human friendly name of your plugin in Processing."""
```
Comment (nyalldawson) [inline source/docs/pyqgis_developer_cookbook/processing.rst]:
```suggestion
           """The human friendly name of your plugin in Processing.
           This string should be as short as possible (e.g. "Lastools", not "Lastools version 1.0.1 64-bit") and localised."""
```

File: source/docs/pyqgis_developer_cookbook/processing.rst [hunk 9]
```diff
@@ -22,18 +20,98 @@ processing interface), and a quicker development time (since Processing will tak
...
+           return 'Your plugin'
+
+       def icon(self):
+           """The QIcon in the Processing toolbox."""
```
Comment (nyalldawson) [inline source/docs/pyqgis_developer_cookbook/processing.rst]:
```suggestion
           """Should return a QIcon which is used for your provider inside the Processing toolbox."""
```

File: source/docs/pyqgis_developer_cookbook/processing.rst (hunk 6 above)
Comment (Gustry) [reply source/docs/pyqgis_developer_cookbook/processing.rst]: I agree. I think it was legacy from QGIS 2.
But I was sure I got a kind of ownership problem `Python error:RuntimeError: wrapped C/C++ object of type XXXX has been deleted`

I will check thanks.

Comment (nyalldawson) [reply source/docs/pyqgis_developer_cookbook/processing.rst]: That only happens when plugins store algorithms and try to reuse them in loadAlgorithms - e.g.

    def __init__(self):
       self.alg1 = MyAlg1()
       self.alg2 = MyAlg2()

    def loadAlgorithms(...):
       self.addAlgorithm(self.alg1)
       self.addAlgorithm(self.alg2)

For some reason this was a common approach in 2.x, but it's NOT valid for 3.x, where addAlgorithm takes ownership of the algorithm and requires a new instance of each algorithm.

File: source/docs/pyqgis_developer_cookbook/processing.rst [hunk 10]
```diff
@@ -22,18 +20,98 @@ processing interface), and a quicker development time (since Processing will tak
...
+
+       def __init__(self):
+           self.provider = None
+
+       def initProcessing(self):
+           self.provider = Provider()
+           QgsApplication.processingRegistry().addProvider(self.provider)
+
+       def initGui(self):
+           if Qgis.QGIS_VERSION_INT < 30800:
```
[the remark this answers is not in the document]
Comment (Gustry) [reply source/docs/pyqgis_developer_cookbook/processing.rst]: Hum, I misunderstood your PR so.
Ok thanks for your explanations.

But do you confirm that dev should use the `initProcessing` name?
So when `initProcessing` will be called by QGIS? In the standalone QGIS Processing executable?

[the remark this answers is not in the document]
Comment (nyalldawson) [reply source/docs/pyqgis_developer_cookbook/processing.rst]: Yes, and for standalone scripts which require load processing plugins

[the remark this answers is not in the document]
Comment (Gustry) [reply source/docs/pyqgis_developer_cookbook/processing.rst]: Ok thanks. I understand now. I really thought that 'initProcessing` was called by QGIS, the same as `initGui`.
And I tried this code and it worked with the QGIS Master. Anyway, I will update.

[the remark this answers is not in the document]
Comment (nyalldawson) [reply source/docs/pyqgis_developer_cookbook/processing.rst]: Well - I'd love to make that so, but I don't think it's possible without breaking API. We'd either be risking not calling it at all, or risk having it called twice...

File: source/docs/pyqgis_developer_cookbook/processing.rst [hunk 11]
```diff
@@ -22,18 +20,100 @@ processing interface), and a quicker development time (since Processing will tak
...
   file and the algorithm file are fully commented and contain information about how to
   modify the provider and add additional algorithms. Refer to them for more information.
 
+If you want to add your existing plugin to Processing, you need to add some code.
+
+In your :file:`metadata.txt`, you need to add a variable:
+
+.. code-block:: ini
+
+    hasProcessingProvider=yes
```
Comment (Gustry) [inline source/docs/pyqgis_developer_cookbook/processing.rst]: @nyalldawson Do you think it will be updated to `True`?

Comment (nyalldawson) [reply source/docs/pyqgis_developer_cookbook/processing.rst]: It's following the same pattern as experimental, which uses "yes"

Comment (Gustry) [reply source/docs/pyqgis_developer_cookbook/processing.rst]: https://github.com/qgis/QGIS/blob/master/src/app/pluginmanager/qgspluginmanager.cpp#L816 and L825.

And MetaSearch https://github.com/qgis/QGIS/blob/master/python/plugins/MetaSearch/metadata.txt#L15

Comment (Gustry) [reply source/docs/pyqgis_developer_cookbook/processing.rst]: I just tried on my own plugin. Both work `yes` and `true` ;-)
We find a mix of both in some plugins.

Comment (nyalldawson) [reply source/docs/pyqgis_developer_cookbook/processing.rst]: Oh good, I'm not going crazy :) I could have sworn I copied this from another plugin somewhere

**Comment by havatv** [comment]
I can't find initProcessing in 3.4.  Should it be backported?

**Comment by Gustry** [comment]
`initProcessing` is coming in QGIS 3.8.
But doing like this, it makes the plugin compatible for all QGIS 3 version (< 3.8), so it's fine.

Dev will make their plugin already designed for QGIS 3.8 without being aware ;-)

**Comment by havatv** [comment]
You are right - there does not seem to be a problem with backporting.

**Comment by ghtmtt** [comment]
ok guys, I'm going to merge this. I think new example is well explained and code has been reviewed ;)

**Comment by ghtmtt** [comment]
thanks @Gustry @havatv and @nyalldawson for the code review

**Pull request merged**